sdk/src/cpi.rs

```diff
@@ -0,0 +1,140 @@
+            data_addr: data.as_ptr() as u64,
+            data_len: data.len(),
+        }
+    }
+}
+
+/// The array elem of `sol_invoke_signed_c()`'s `account_infos_addr` arg.
+///
+/// The u64 fields are raw pointers.
+///
+/// DO NOT EXPOSE THIS STRUCT -
+/// to ensure pointers are valid upon use, the scope of this struct should
+/// only be limited to the stack where sol_invoke_signed_c happens and then
+/// discarded immediately after
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+struct SolAccountInfo {
+    key_addr: u64,
+    lamports_addr: u64,
+    data_len: u64,
+    data_addr: u64,
+    owner_addr: u64,
+    rent_epoch: u64,
+    is_signer: bool,
+    is_writable: bool,
+    executable: bool,
+}
+
+impl SolAccountInfo {
+    const fn null() -> Self {
+        Self {
+            key_addr: 0,
+            lamports_addr: 0,
+            data_len: 0,
+            data_addr: 0,
+            owner_addr: 0,
+            rent_epoch: 0,
+            is_signer: false,
+            is_writable: false,
+            executable: false,
+        }
+    }
+}
+
+impl From<&AccountInfo> for SolAccountInfo {
+    #[inline]
+    fn from(value: &AccountInfo) -> Self {
+        let (data_len, data_addr) = {
+            let a = unsafe { value.unchecked_borrow_data() };
+            let len = a.len();
+            (len, a.as_ptr())
+        };
+        let lamports_addr = unsafe { value.unchecked_borrow_lamports() } as *const _;
+        Self {
+            key_addr: value.key().as_ptr() as u64,
+            lamports_addr: lamports_addr as u64,
+            data_len: data_len as u64,
+            data_addr: data_addr as u64,
+            owner_addr: value.owner().as_ptr() as u64,
+            is_signer: value.is_signer(),
+            is_writable: value.is_writable(),
+            executable: value.executable(),
+            rent_epoch: 0, // dont care
+        }
+    }
+}
+
+pub fn invoke_signed<'a, const MAX_ACCOUNTS_INCL_PROGRAM: usize>(
+    ix: Instruction,
+    program_account_info: &'a AccountInfo,
+    account_infos: impl IntoIterator<Item = &'a AccountInfo>,
+    signers: &[&[&[u8]]],
+) -> Result<(), ProgramError> {
+    let (sol_account_infos, sol_account_infos_len) = core::iter::once(program_account_info)
+        .chain(account_infos)
+        .enumerate()
+        .try_fold(
+            ([SolAccountInfo::null(); MAX_ACCOUNTS_INCL_PROGRAM], 0),
+            |(mut arr, len), (i, curr)| {
+                if i >= MAX_ACCOUNTS_INCL_PROGRAM {
+                    return Err(ProgramError::InvalidArgument);
+                }
+                arr[i] = SolAccountInfo::from(curr);
+                Ok((arr, len + 1))
+            },
+        )?;
+    let sol_ix = SolInstruction::from(ix);
+    let result = unsafe {
+        sol_invoke_signed_c(
+            &sol_ix as *const _ as *const _,
+            &sol_account_infos as *const _ as *const _,
+            sol_account_infos_len,
+            signers as *const _ as *const _,
```

Comment:
We can't pass the rust references directly to the C code, since it expects the length of each individual seed unfortunately. I started adding signer "types" on this PR: https://github.com/febo/pinocchio/pull/11/files#diff-c14b0096da794786d77f37e2667c1f5cf800e7bf7875f8e99059e14857a03cf4R107-R127

Reply:
For reference: https://github.com/anza-xyz/agave/blob/master/programs/bpf_loader/src/syscalls/cpi.rs#L596-L610